core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PushMergeWriterUpdateIntoConnector.java

```diff
@@ -0,0 +1,134 @@
+import static io.trino.matching.Capture.newCapture;
+import static io.trino.sql.planner.plan.Patterns.mergeProcessor;
+import static io.trino.sql.planner.plan.Patterns.mergeWriter;
+import static io.trino.sql.planner.plan.Patterns.project;
+import static io.trino.sql.planner.plan.Patterns.source;
+import static io.trino.sql.planner.plan.Patterns.tableFinish;
+import static io.trino.sql.planner.plan.Patterns.tableScan;
+import static java.util.Objects.requireNonNull;
+
+public class PushMergeWriterUpdateIntoConnector
+        implements Rule<TableFinishNode>
+{
+    private static final Capture<TableScanNode> TABLE_SCAN = newCapture();
+    private static final Capture<ProjectNode> PROJECT_NODE_CAPTURE = newCapture();
+    private static final Capture<MergeWriterNode> MERGE_WRITER_NODE_CAPTURE = newCapture();
+
+    private static final Pattern<TableFinishNode> PATTERN =
+            tableFinish().with(source().matching(
+                    mergeWriter().capturedAs(MERGE_WRITER_NODE_CAPTURE).with(source().matching(
+                            mergeProcessor().with(source().matching(
+                                    project().with(source().matching(
+                                            tableScan().capturedAs(TABLE_SCAN))).capturedAs(PROJECT_NODE_CAPTURE)))))));
+
+    public PushMergeWriterUpdateIntoConnector(PlannerContext plannerContext, TypeAnalyzer typeAnalyzer, Metadata metadata)
+    {
+        this.metadata = requireNonNull(metadata, "metadata is null");
+        this.typeAnalyzer = requireNonNull(typeAnalyzer, "typeAnalyzer is null");
+        this.plannerContext = requireNonNull(plannerContext, "plannerContext is null");
+    }
+
+    private final Metadata metadata;
+    private final PlannerContext plannerContext;
+    private final TypeAnalyzer typeAnalyzer;
+
+    @Override
+    public Pattern<TableFinishNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(TableFinishNode node, Captures captures, Context context)
+    {
+        TableScanNode tableScan = captures.get(TABLE_SCAN);
+        ProjectNode project = captures.get(PROJECT_NODE_CAPTURE);
+        MergeWriterNode mergeWriter = captures.get(MERGE_WRITER_NODE_CAPTURE);
+
+        List<String> orderedColumnNames = mergeWriter.getTarget().getMergeParadigmAndTypes().getColumnNames();
+        List<? extends Node> mergeAssignments = project.getAssignments().get(new Symbol("merge_row")).getChildren();
+
+        ImmutableMap.Builder<String, Domain> assignmentsBuilder = ImmutableMap.builder();
+
+        for (int i = 0; i < orderedColumnNames.size(); i++) {
+            String columnName = orderedColumnNames.get(i);
+            Node assigmentNode = mergeAssignments.get(i);
+
+            // we don't support yet any expressions in update statements, only constants
+            // also in case of UPDATE ALL statement, optimiser transforms expression to base64 call
```

Comment:
As far as I see in debug usually merge row looks like this:
- it's kind of list of values in column order, where we have either value (which we want to set) or column name (it means colum stay untouched), like
` 1, colum_date, 'String value'`, e.t.c
But in case we perform update of ALL values of columns (every column is updated), this list of values is encoded in some base64 string by optimiser.
And in this case instead of usual format (list of column values), which is easy to parse -  we got some encoded Base64 String, which require some additional parsing to extract the update values.
I plan to add this functionality next, but for now I decided just to skip such cases.

Reply:
Sure,
for such query, when we update all columns in table:
```
        try (TestTable table = new TestTable(getQueryRunner()::execute, "test_update_all_columns", "(a INT, b INT, c INT)")) {
            String tableName = table.getName();
            assertUpdate("INSERT INTO " + tableName + " VALUES (1, 2, 3), (11, 12, 13), (21, 22, 23)", 3);
            assertUpdate("UPDATE " + tableName + " SET a = 10, b = 20, c = 20", 3);
        }
```

merger row looks like this:
<img width="2139" alt="base64" src="https://github.com/trinodb/trino/assets/112624963/bc730cec-14d9-45b5-bdcd-633aeddee335">

Reply:
Ok, I see. That is just how a value of `row` type is being encoded when all the fields of the row are constant, which happens when all the columns are set. Using base64 is an internal implementation detail that's irrelevant to this optimizer.